File: 2019/day23.py
import queue
import threading

from utils import read_data
from utils.intcode import VM


class NetworkedVM:
    def __init__(self, program, address, network):
        self.address = address
        self.vm = VM(program, inputs=self.inputs())
        self.ready = threading.Event()
        self.queue = queue.Queue()
        self.network = network
        self.add_message((address, ))
        self.running = False

    def inputs(self):
        while self.running:
            self.ready.wait()
            try:
                values = self.queue.get(block=False)
            except queue.Empty:
                values = (-1, )
                self.ready.clear()
                self.network.nat.notify_idle()
            yield from values

    def add_message(self, message):
        self.queue.put(message)
        self.ready.set()

    def run(self):
        self.running = True
        outputs = self.vm.outputs()
        while self.running:
            try:
                destination = next(outputs)
                x = next(outputs)
                y = next(outputs)
            except StopIteration:
                self.running = False  # Was Network!
                break
            self.network.send_message(self.address, destination, (x, y))

    def stop(self):
        self.running = False
        self.ready.set()


class NAT:
    def __init__(self, network):
        self.network = network
        self.condition = threading.Condition()
        self.last_message = None
        self.running = False

    def notify_idle(self):
        with self.condition:
            self.condition.notify()

    def add_message(self, message):
        self.last_message = message

    def run(self):
        def all_vms_idle():
            idle_vms = [vm for vm in self.network.vms if not vm.ready.is_set()]
            return len(idle_vms) == len(self.network.vms)

        self.running = True
        while self.running:
            with self.condition:
                self.condition.wait_for(lambda: all_vms_idle() or not self.running)
            self.network.send_message(255, 0, self.last_message)

    def stop(self):
        self.running = False
        with self.condition:
            self.condition.notify()


class Network:
    def __init__(self, num_machines, program):
        self.vms = []
        self.nat = NAT(self)
        self.messages_sent = queue.Queue()
        for network_address in range(num_machines):
            self.vms.append(NetworkedVM(program[:], network_address, self))

    def send_message(self, source, destination, message):
        if not 0 <= destination < len(self.vms) and not destination == 255:
            raise RuntimeError(f"Invalid destination: {destination}")
        vm = self.nat if destination == 255 else self.vms[destination]
        self.messages_sent.put((source, destination, message))
        vm.add_message(message)

    def start(self):
        thread = threading.Thread(target=self.nat.run)
        thread.start()
        for vm in self.vms:
            thread = threading.Thread(target=vm.run)
            thread.start()

    def stop(self):
        self.nat.stop()
        for vm in self.vms:
            vm.stop()


def p1(program):
    network = Network(50, program)
    network.start()
    destination = None
    while destination != 255:
        source, destination, (x, y) = network.messages_sent.get()
    network.stop()
    return y


def p2(program):
    network = Network(50, program)
    network.start()
    last_y = None
    while True:
        source, destination, (x, y) = network.messages_sent.get()
        if (source, destination) == (255, 0):
            if y == last_y:
                break
            last_y = y
    network.stop()
    return y


if __name__ == "__main__":
    print("No tests run")

    data = read_data(23)
    program = [int(x) for x in data.strip().split(',')]
    print("Part 1: {}".format(p1(program)))
    print("Part 2: {}".format(p2(program)))

File: 2019/rust/src/bin/day23.rs
#![feature(async_closure)]

use std::sync::Arc;

use anyhow::{anyhow, Result};
use futures::stream::{Stream, StreamExt};
use tokio::sync::{mpsc, Notify};

use aoc::intcode::{RegisterValue, VM};

type Address = usize;

#[tokio::main]
async fn main() -> Result<()> {
    env_logger::init();
    let input = aoc::utils::get_input()?;
    let program: Vec<RegisterValue> = aoc::intcode::parse_program(input.trim())?;

    println!("Part 1: {}", part1(&program).await?);

    Ok(())
}

async fn part1(program: &[RegisterValue]) -> Result<RegisterValue> {
    let mut network = Network::start(program, 50);
    while let Some(message) = network.incoming.recv().await {
        println!("Processing {:?}", message);
        if message.destination == 255 {
            return Ok(message.data.y);
        }
        let channel = network
            .vm_channels
            .get(message.destination)
            .ok_or_else(|| anyhow!("Channel not found for {}.", message.destination))?;
        channel.send(message.data)?;
    }
    Err(anyhow!("Network terminated early"))
}

struct Network {
    vm_channels: Vec<NetworkedVMChannel>,
    incoming: mpsc::UnboundedReceiver<Message>,
}

impl Network {
    fn start(program: &[RegisterValue], number: Address) -> Network {
        let (outputs, incoming) = mpsc::unbounded_channel();
        let mut vm_channels = vec![];
        for address in 0..number {
            vm_channels.push(NetworkedVMChannel::spawn(program, address, outputs.clone()));
        }
        vm_channels[0].ready.notify_one();
        Network {
            vm_channels,
            incoming,
        }
    }
}

struct NetworkedVMChannel {
    channel: mpsc::UnboundedSender<Packet>,
    ready: Arc<Notify>,
}

impl NetworkedVMChannel {
    fn spawn(
        program: &[RegisterValue],
        address: Address,
        outputs: mpsc::UnboundedSender<Message>,
    ) -> NetworkedVMChannel {
        let (channel, inputs) = mpsc::unbounded_channel();
        let ready = Arc::new(Notify::new());
        let mut vm = NetworkedVM::new(program.to_vec(), address, inputs, outputs, ready.clone());
        tokio::spawn(async move { vm.run().await });
        NetworkedVMChannel { channel, ready }
    }

    fn send(&self, packet: Packet) -> Result<()> {
        self.channel.send(packet)?;
        self.ready.notify_one();
        Ok(())
    }
}

struct NetworkedVM<'a> {
    vm: VM<'a>,
    address: Address,
    channel: mpsc::UnboundedSender<Message>,
}

impl<'a> NetworkedVM<'a> {
    fn new(
        program: Vec<RegisterValue>,
        address: Address,
        inputs: mpsc::UnboundedReceiver<Packet>,
        channel: mpsc::UnboundedSender<Message>,
        ready: Arc<Notify>,
    ) -> NetworkedVM<'a> {
        let vm = VM::new(program.to_vec(), Box::pin(input_stream(inputs, ready)));
        NetworkedVM {
            vm,
            address,
            channel,
        }
    }

    async fn run(&mut self) -> Option<()> {
        let mut outputs = self.vm.outputs_async();
        loop {
            let destination = outputs.next().await?;
            let x = outputs.next().await?;
            let y = outputs.next().await?;
            let message = Message {
                source: self.address,
                destination: destination as usize, // TODO: handle error
                data: Packet { x, y },
            };
            println!("Sending {:?}", message);
            self.channel.send(message).unwrap(); // TODO
            tokio::task::yield_now().await; // TODO
        }
    }
}

#[derive(Debug, Clone, Copy)]
struct Message {
    source: Address,
    destination: Address,
    data: Packet,
}

#[derive(Debug, Clone, Copy)]
struct Packet {
    x: RegisterValue,
    y: RegisterValue,
}

fn input_stream(
    channel: mpsc::UnboundedReceiver<Packet>,
    ready: Arc<Notify>,
) -> impl Stream<Item = RegisterValue> {
    futures::stream::unfold(
        (channel, ready, None),
        async move |(mut channel, ready, next_value)| {
            if let Some(value) = next_value {
                return Some((value, (channel, ready, None)));
            }
            ready.notified().await;
            println!("Received Notification");
            match channel.try_recv() {
                Ok(message) => {
                    println!("Got {:?}", message);
                    Some((message.x, (channel, ready, Some(message.y))))
                }
                Err(mpsc::error::TryRecvError::Empty) => Some((-1, (channel, ready, None))),
                Err(mpsc::error::TryRecvError::Disconnected) => None,
            }
        },
    )
}
